packages/user-operation-controller/src/utils/gas-fees.ts

```diff
@@ -0,0 +1,295 @@
+ * @returns The maxPriorityFeePerGas for the user operation.
+ */
+async function getMaxPriorityFeePerGas(
+  originalRequest: AddUserOperationRequest,
+  getGetFasEstimates: () => Promise<SuggestedGasFees>,
+  userOperation: UserOperation,
+  transaction?: TransactionParams,
+) {
+  const { maxFeePerGas, maxPriorityFeePerGas } = originalRequest;
+  const { gasPrice } = transaction ?? {};
+  const { maxFeePerGas: updatedMaxFeePerGas } = userOperation;
+
+  if (!isGasFeeEmpty(maxPriorityFeePerGas)) {
+    log('Using maxPriorityFeePerGas from request', maxPriorityFeePerGas);
+    return maxPriorityFeePerGas as string;
+  }
+
+  if (isGasFeeEmpty(maxFeePerGas) && gasPrice) {
+    log('Setting maxPriorityFeePerGas to transaction gasPrice', gasPrice);
+    return gasPrice;
+  }
+
+  const { maxPriorityFeePerGas: suggestedMaxPriorityFeePerGas } =
+    await getGetFasEstimates();
+
+  if (suggestedMaxPriorityFeePerGas) {
+    log(
+      'Using maxPriorityFeePerGas from estimate',
+      suggestedMaxPriorityFeePerGas,
+    );
+    return suggestedMaxPriorityFeePerGas;
+  }
+
+  log('Setting maxPriorityFeePerGas to maxFeePerGas', updatedMaxFeePerGas);
+
+  return updatedMaxFeePerGas;
+}
+
+/**
+ * Gets the userFeeLevel for a user operation.
+ * @param metadata - The metadata for the user operation.
+ * @param originalRequest - The original request to add the user operation.
+ * @param suggestedGasFees - The suggested gas fees, if any.
+ * @param transaction - The transaction that created the user operation.
+ * @returns The userFeeLevel for the user operation.
+ */
+function getUserFeeLevel(
+  metadata: UserOperationMetadata,
+  originalRequest: AddUserOperationRequest,
+  suggestedGasFees?: SuggestedGasFees,
+  transaction?: TransactionParams,
+): UserFeeLevel {
+  const { origin } = metadata;
+  const { maxFeePerGas, maxPriorityFeePerGas } = originalRequest;
+
+  const {
+    maxFeePerGas: suggestedMaxFeePerGas,
+    maxPriorityFeePerGas: suggestedMaxPriorityFeePerGas,
+  } = suggestedGasFees || {};
+
+  if (
+    isGasFeeEmpty(maxFeePerGas) &&
+    isGasFeeEmpty(maxPriorityFeePerGas) &&
+    transaction?.gasPrice
+  ) {
+    return origin === ORIGIN_METAMASK
+      ? UserFeeLevel.CUSTOM
+      : UserFeeLevel.DAPP_SUGGESTED;
+  }
+
+  if (
+    isGasFeeEmpty(maxFeePerGas) &&
+    isGasFeeEmpty(maxPriorityFeePerGas) &&
+    suggestedMaxFeePerGas &&
+    suggestedMaxPriorityFeePerGas
+  ) {
+    return UserFeeLevel.MEDIUM;
+  }
+
+  if (origin === ORIGIN_METAMASK) {
+    return UserFeeLevel.CUSTOM;
+  }
+
+  return UserFeeLevel.DAPP_SUGGESTED;
+}
+
+/**
+ * Gets suggested gas fees.
+ * @param request - The request to update the gas fees.
+ * @param request.getGasFeeEstimates - A callback to get gas fee estimates.
+ * @param request.provider - A provider to query the network.
+ * @returns The suggested gas fees.
+ */
+async function getSuggestedGasFees(request: UpdateGasFeesRequest) {
```

Comment:
Matt, could we add the return type here? Something like:

Promise<{
    maxFeePerGas?: Hex;
    maxPriorityFeePerGas?: Hex;
}>

Reply:
Done.